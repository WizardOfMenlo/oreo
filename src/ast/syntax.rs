//! Typed syntax that is backed by Node

use super::node_db::{NodeDb, NodeId};
use super::untyped::{Node, NodeType};

/// Macro for simple definition of a syntax node
macro_rules! syntax_node {
    ($id:ident) => {
        /// A syntax node for an $id
        #[derive(Debug, Clone, Copy, Hash, Eq, PartialEq)]
        pub struct $id(NodeId);

        impl $id {
            /// Instantiate the syntax node
            pub fn new(node: NodeId) -> Self {
                $id(node)
            }

            /// Get node id
            pub fn get_id(self) -> NodeId {
                self.0
            }

            /// Get the raw node
            pub fn get_node<'a, 'b>(self, db: &'b NodeDb<'a>) -> &'b Node<'a> {
                db.get_node(self.0).expect("Invalid syntax node")
            }

            /// Get children
            pub fn children<'a, 'b>(self, db: &'b NodeDb<'a>) -> &'b [NodeId] {
                db.get_flat_node(self.0)
                    .expect("Invalid syntax node")
                    .children()
            }
        }
    };
}

syntax_node!(Program);

impl Program {
    /// Get the program name
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// Program main body
    pub fn compound<'a>(self, db: &NodeDb<'a>) -> Compound {
        Compound(self.children(db)[1])
    }
}

syntax_node!(Compound);

impl Compound {
    /// Get the inner statements
    pub fn statements<'a>(self, db: &NodeDb<'a>) -> Vec<Statement> {
        self.children(db).iter().map(|n| Statement(*n)).collect()
    }
}

syntax_node!(Statement);

impl Statement {
    /// Convert the statement to a more specific one
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> StatementType {
        let n = self.0;
        match self.get_node(db).ty() {
            NodeType::Decl => StatementType::Decl(Decl(n)),
            NodeType::PrintStat => StatementType::PrintStat(PrintStat(n)),
            NodeType::While => StatementType::While(While(n)),
            NodeType::If => StatementType::If(If(n)),
            NodeType::Assign => StatementType::Assign(Assign(n)),
            NodeType::FunctionCall => StatementType::FunctionCall(FunctionCall(n)),
            NodeType::FunctionDecl => StatementType::FunctionDecl(FunctionDecl(n)),
            NodeType::Return => StatementType::Return(Return(n)),
            _ => panic!("Invalid node constructed"),
        }
    }
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
/// The possible types that a statement can be
pub enum StatementType {
    Decl(Decl),
    PrintStat(PrintStat),
    While(While),
    If(If),
    Assign(Assign),
    FunctionCall(FunctionCall),
    FunctionDecl(FunctionDecl),
    Return(Return),
}

syntax_node!(Decl);

impl Decl {
    /// The identifier being declared
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The expression to assign (if any)
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Option<Expr> {
        self.children(db).get(1).map(|n| Expr(*n))
    }
}

syntax_node!(PrintStat);

impl PrintStat {
    /// Convert to the more specific print typ
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> PrintTypes {
        let node = self.children(db)[0];
        match self.get_node(db).ty() {
            NodeType::Print => PrintTypes::Print(Print(node)),
            NodeType::Println => PrintTypes::Println(Println(node)),
            NodeType::Get => PrintTypes::Get(Get(node)),
            _ => panic!("Invalid node constructed"),
        }
    }
}

/// Possible print types
#[derive(Debug, Clone)]
#[allow(missing_docs)]
pub enum PrintTypes {
    Print(Print),
    Println(Println),
    Get(Get),
}

syntax_node!(Print);

impl Print {
    /// The expression to print
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(Println);

impl Println {
    /// The expression to print
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(Get);

impl Get {
    /// The identifier to set
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }
}

syntax_node!(While);

impl While {
    /// The condition to evaluate
    pub fn condition<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }

    /// The body to be executed
    pub fn compound<'a>(self, db: &NodeDb<'a>) -> Compound {
        Compound(self.children(db)[1])
    }
}

syntax_node!(If);

impl If {
    /// The condition to evaluate
    pub fn condition<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }

    /// If true
    pub fn if_branch<'a>(self, db: &NodeDb<'a>) -> Compound {
        Compound(self.children(db)[1])
    }

    /// If false (optional)
    pub fn else_branch<'a>(self, db: &NodeDb<'a>) -> Option<Compound> {
        self.children(db).get(2).map(|n| Compound(*n))
    }
}

syntax_node!(Assign);

impl Assign {
    /// The identifier to set
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The expression to set it to
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[1])
    }
}

syntax_node!(FunctionDecl);

impl FunctionDecl {
    /// Function to declare
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The arguments
    pub fn args<'a>(self, db: &NodeDb<'a>) -> Vec<FunctionDeclArgs> {
        let children = self.children(db);
        children[1..children.len() - 1]
            .iter()
            .map(|n| FunctionDeclArgs(*n))
            .collect()
    }

    /// The body to be executed
    pub fn compound<'a>(self, db: &NodeDb<'a>) -> Compound {
        let children = self.children(db);
        Compound(children[children.len() - 1])
    }
}

syntax_node!(FunctionDeclArgs);

impl FunctionDeclArgs {
    /// The arg id
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }
}

syntax_node!(FunctionCallArgs);

impl FunctionCallArgs {
    /// The expression in the argument
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(Return);

impl Return {
    /// The expression to return
    pub fn expr<'a>(self, db: &NodeDb<'a>) -> Expr {
        Expr(self.children(db)[0])
    }
}

syntax_node!(FunctionCall);

impl FunctionCall {
    /// The function id to call
    pub fn id<'a>(self, db: &NodeDb<'a>) -> Identifier {
        Identifier(self.children(db)[0])
    }

    /// The arguments of the call
    pub fn args<'a>(self, db: &NodeDb<'a>) -> Vec<FunctionCallArgs> {
        self.children(db)[1..]
            .iter()
            .map(|n| FunctionCallArgs(*n))
            .collect()
    }
}

syntax_node!(Expr);

impl Expr {
    /// The head of the expression
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Term {
        Term(self.children(db)[0])
    }

    /// Tail of the expression
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ExprPrime> {
        self.children(db).get(1).map(|e| ExprPrime(*e))
    }
}

syntax_node!(ExprPrime);

impl ExprPrime {
    /// The operation
    pub fn op<'a>(self, db: &NodeDb<'a>) -> BooleanOp {
        self.get_node(db).ty().clone().into()
    }

    /// The right operand
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Term {
        Term(self.children(db)[0])
    }

    /// Possible tails
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ExprPrime> {
        self.children(db).get(1).map(|e| ExprPrime(*e))
    }
}

syntax_node!(Term);

impl Term {
    /// Head of the term
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Factor {
        Factor(self.children(db)[0])
    }

    /// Tail of the term
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<TermPrime> {
        self.children(db).get(1).map(|n| TermPrime(*n))
    }
}

syntax_node!(TermPrime);

impl TermPrime {
    /// The operation
    pub fn op<'a>(self, db: &NodeDb<'a>) -> RelationalOp {
        self.get_node(db).ty().clone().into()
    }

    /// Rhs
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Factor {
        Factor(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<TermPrime> {
        self.children(db).get(1).map(|n| TermPrime(*n))
    }
}

syntax_node!(Factor);

impl Factor {
    /// The head of a factor
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Product {
        Product(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<FactorPrime> {
        self.children(db).get(1).map(|n| FactorPrime(*n))
    }
}

syntax_node!(FactorPrime);

impl FactorPrime {
    /// The operation in the tail
    pub fn op<'a>(self, db: &NodeDb<'a>) -> AdditiveOp {
        self.get_node(db).ty().clone().into()
    }

    /// The rhs of the tail
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Product {
        Product(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<FactorPrime> {
        self.children(db).get(1).map(|n| FactorPrime(*n))
    }
}

syntax_node!(Product);

impl Product {
    /// The head
    pub fn head<'a>(self, db: &NodeDb<'a>) -> Atom {
        Atom(self.children(db)[0])
    }

    /// Opt tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ProductPrime> {
        self.children(db).get(1).map(|n| ProductPrime(*n))
    }
}

syntax_node!(ProductPrime);

impl ProductPrime {
    /// The operation
    pub fn op<'a>(self, db: &NodeDb<'a>) -> MultiplicativeOp {
        self.get_node(db).ty().clone().into()
    }

    /// The rhs of the product
    pub fn rhs<'a>(self, db: &NodeDb<'a>) -> Atom {
        Atom(self.children(db)[0])
    }

    /// Optional tail
    pub fn tail<'a>(self, db: &NodeDb<'a>) -> Option<ProductPrime> {
        self.children(db).get(1).map(|n| ProductPrime(*n))
    }
}

syntax_node!(Atom);

/// Possible types of an atom
pub enum AtomType {
    /// Literal, function call or bracketed expr
    Unit(Unit),

    /// The not of an atom
    Not(Atom),
}

impl Atom {
    /// Convert an atom to a more specific type
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> AtomType {
        let n = self.children(db)[0];
        match self.get_node(db).ty() {
            NodeType::Not => AtomType::Not(Atom(n)),
            NodeType::Unit => AtomType::Unit(Unit(n)),
            _ => panic!("Invalid atom type"),
        }
    }
}

syntax_node!(Unit);

impl Unit {
    /// Convert a unit into a more specific type
    pub fn downcast<'a>(self, db: &NodeDb<'a>) -> UnitType {
        match self.get_node(db).ty() {
            NodeType::Int(i) => UnitType::Int(*i),
            NodeType::Bool(b) => UnitType::Bool(*b),
            NodeType::Str => UnitType::Str(self.0),
            NodeType::Identifier => UnitType::Identifier(Identifier(self.children(db)[0])),
            NodeType::FunctionCall => UnitType::FunctionCall(FunctionCall(self.children(db)[0])),
            NodeType::BracketedExpr => UnitType::BracketedExpr(Expr(self.children(db)[0])),
            _ => panic!("Invalid atom type"),
        }
    }
}

/// Possible unit types
#[derive(Debug, Clone)]
#[allow(missing_docs)]
pub enum UnitType {
    Int(isize),
    Str(NodeId),
    Bool(bool),
    Identifier(Identifier),
    FunctionCall(FunctionCall),
    BracketedExpr(Expr),
}

syntax_node!(Identifier);

impl Identifier {
    /// Get the string id
    pub fn id<'a>(self, db: &NodeDb<'a>, text: &'a str) -> &'a str {
        use crate::range::Ranged;
        &text[self.get_node(db).range().clone()]
    }
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum AdditiveOp {
    Plus,
    Minus,
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum MultiplicativeOp {
    Times,
    Divide,
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum RelationalOp {
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,
}

#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum BooleanOp {
    And,
    Or,
}

/// Enum to convert any node into a typed one
#[allow(missing_docs)]
#[derive(Debug, Clone)]
pub enum SyntaxNode {
    Program(Program),
    Compound(Compound),
    Decl(Decl),
    PrintStat(PrintStat),
    Print(Print),
    Println(Println),
    Get(Get),
    While(While),
    If(If),
    Assign(Assign),
    FunctionDecl(FunctionDecl),
    FunctionDeclArgs(FunctionDeclArgs),
    FunctionCall(FunctionCall),
    FunctionCallArgs(FunctionCallArgs),
    Return(Return),
    Expr(Expr),
    ExprPrime(ExprPrime),
    Term(Term),
    TermPrime(TermPrime),
    Factor(Factor),
    FactorPrime(FactorPrime),
    Product(Product),
    ProductPrime(ProductPrime),
    Atom(Atom),
    Unit(Unit),
    Identifier(Identifier),
}

impl SyntaxNode {
    /// Convert a id to a syntax node
    pub fn from<'a>(input: NodeId, db: &NodeDb<'a>) -> Self {
        match db.get_node(input).expect("Invalid id").ty() {
            NodeType::Program => SyntaxNode::Program(Program(input)),
            NodeType::Compound => SyntaxNode::Compound(Compound(input)),
            NodeType::Decl => SyntaxNode::Decl(Decl(input)),
            NodeType::PrintStat => SyntaxNode::PrintStat(PrintStat(input)),

            NodeType::While => SyntaxNode::While(While(input)),
            NodeType::If => SyntaxNode::If(If(input)),
            NodeType::Assign => SyntaxNode::Assign(Assign(input)),
            NodeType::FunctionCall => SyntaxNode::FunctionCall(FunctionCall(input)),
            NodeType::FunctionDecl => SyntaxNode::FunctionDecl(FunctionDecl(input)),
            NodeType::Return => SyntaxNode::Return(Return(input)),
            NodeType::Print => SyntaxNode::Print(Print(input)),
            NodeType::Println => SyntaxNode::Println(Println(input)),
            NodeType::Get => SyntaxNode::Get(Get(input)),
            NodeType::FunctionDeclArgs => SyntaxNode::FunctionDeclArgs(FunctionDeclArgs(input)),
            NodeType::FunctionCallArgs => SyntaxNode::FunctionCallArgs(FunctionCallArgs(input)),

            NodeType::Expr => SyntaxNode::Expr(Expr(input)),
            NodeType::And | NodeType::Or => SyntaxNode::ExprPrime(ExprPrime(input)),

            NodeType::Term => SyntaxNode::Term(Term(input)),
            NodeType::LesserThan
            | NodeType::GreaterThan
            | NodeType::Equals
            | NodeType::GreaterOrEquals
            | NodeType::LesserOrEquals => SyntaxNode::TermPrime(TermPrime(input)),

            NodeType::Factor => SyntaxNode::Factor(Factor(input)),
            NodeType::Plus | NodeType::Minus => SyntaxNode::FactorPrime(FactorPrime(input)),

            NodeType::Product => SyntaxNode::Product(Product(input)),
            NodeType::Times | NodeType::Divide => SyntaxNode::ProductPrime(ProductPrime(input)),

            NodeType::Unit | NodeType::Not => SyntaxNode::Atom(Atom(input)),

            NodeType::Int(_) | NodeType::Str | NodeType::Bool(_) | NodeType::BracketedExpr => {
                SyntaxNode::Unit(Unit(input))
            }

            NodeType::Identifier => SyntaxNode::Identifier(Identifier(input)),

            NodeType::Error(_) => panic!("Called on error node"),
        }
    }
}
