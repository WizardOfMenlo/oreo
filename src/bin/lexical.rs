use oreo::lexical::lexicalize;
use oreo::scanner::scan;
use structopt::StructOpt;

#[derive(Debug, StructOpt)]
struct Args {
    /// The input to parse
    #[structopt(short, long)]
    input: String,
}

fn main() {
    let opt = Args::from_args();
    println!("{:?}", lexicalize(scan(&opt.input)).collect::<Vec<_>>());
}
