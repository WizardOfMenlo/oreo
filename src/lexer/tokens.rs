//! Definition of all the tokens that we use

use super::error::LexicalError;
use serde::Serialize;

/// The main token type that we work on
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Token<'a> {
    /// Any of the keywords (apart from and and or)
    Keyword(Keyword),

    /// (, ), ;, ,
    Punctuation(Punctuation),

    /// Addition, mult, ...
    Operator(Operator),

    /// A constant
    Literal(Literal<'a>),

    /// An identifier
    Identifier(&'a str),

    /// A comment
    Comment(&'a str),

    /// Any kind of error we might encounter
    Error(LexicalError<'a>),
}

impl<'a> Token<'a> {
    /// Is the current token an error?
    pub fn is_error(&self) -> bool {
        match self {
            Token::Error(_) => true,
            _ => false,
        }
    }

    /// Is the current token a comment?
    pub fn is_comment(&self) -> bool {
        match self {
            Token::Comment(_) => true,
            _ => false,
        }
    }

    /// Is this token the same as an other (where Int(1) == Int(2))
    pub fn same_kind<'b>(&self, other: &Token<'b>) -> bool
    where
        'b: 'a,
    {
        // This should match every singleton
        if self == other {
            true
        } else {
            // Match literals
            match (self, other) {
                (Self::Literal(Literal::Integer(_)), Self::Literal(Literal::Integer(_))) => true,
                (Self::Literal(Literal::Boolean(_)), Self::Literal(Literal::Boolean(_))) => true,
                (Self::Literal(Literal::String(_)), Self::Literal(Literal::String(_))) => true,
                (Self::Identifier(_), Self::Identifier(_)) => true,
                (Self::Comment(_), Self::Comment(_)) => true,
                _ => false,
            }
        }
    }
}

/// The Keywords we have
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Keyword {
    Program,
    Begin,
    End,
    Var,
    Print,
    Println,
    Get,
    While,
    If,
    Then,
    Else,
    Procedure,
    Return,
}

/// Possible Punctuation marks
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Punctuation {
    Comma,
    Semicolon,
    BracketOpen,
    BracketClose,
}

/// All the operators (including boolean ones)
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Operator {
    Assignement,
    Plus,
    Minus,
    Times,
    Divide,
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,
    Or,
    And,
    Not,
}

/// Constants
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Literal<'a> {
    Integer(isize),
    Boolean(bool),
    String(&'a str),
}
