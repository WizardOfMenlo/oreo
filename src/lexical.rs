use crate::scanner::{Scanned, ScannedItem};
use crate::tokens::{Keyword, LexicalError, Literal, Operator, Punctuation, Token};

use lazy_static::lazy_static;
use std::collections::HashMap;

lazy_static! {
    static ref SINGLE_CHAR_TOKEN: HashMap<char, Token<'static>> = {
        let mut m = HashMap::new();
        m.insert('(', Token::Punctuation(Punctuation::BracketOpen));
        m.insert(')', Token::Punctuation(Punctuation::BracketClose));
        m.insert(';', Token::Punctuation(Punctuation::Semicolon));

        m.insert('+', Token::Operator(Operator::Plus));
        m.insert('-', Token::Operator(Operator::Minus));
        m.insert('*', Token::Operator(Operator::Times));
        m.insert('/', Token::Operator(Operator::Divide));

        m
    };
    static ref KEYWORDS: HashMap<&'static str, Token<'static>> = {
        let mut m = HashMap::new();
        m.insert("program", Token::Keyword(Keyword::Program));
        m.insert("begin", Token::Keyword(Keyword::Begin));
        m.insert("end", Token::Keyword(Keyword::End));
        m.insert("var", Token::Keyword(Keyword::Var));
        m.insert("print", Token::Keyword(Keyword::Print));
        m.insert("println", Token::Keyword(Keyword::Println));
        m.insert("get", Token::Keyword(Keyword::Get));
        m.insert("while", Token::Keyword(Keyword::While));
        m.insert("if", Token::Keyword(Keyword::If));
        m.insert("then", Token::Keyword(Keyword::Then));
        m.insert("else", Token::Keyword(Keyword::Else));

        m.insert("or", Token::Operator(Operator::Or));
        m.insert("and", Token::Operator(Operator::And));
        m.insert("not", Token::Operator(Operator::Not));
        m
    };
}

pub fn lexicalize<'a>(it: impl Iterator<Item = Scanned<'a>>) -> impl Iterator<Item = Token<'a>> {
    it.map(lexicalize_one).flatten()
}

fn lexicalize_one<'a>(scan: Scanned<'a>) -> impl Iterator<Item = Token<'a>> {
    LexicalIt::new(scan)
}

struct LexicalIt<'a> {
    s: Scanned<'a>,
    curr_position: usize,
    terminate: bool,
}

impl<'a> LexicalIt<'a> {
    fn new(s: Scanned<'a>) -> Self {
        LexicalIt {
            s,
            curr_position: 0,
            terminate: false,
        }
    }
}

impl<'a> Iterator for LexicalIt<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.terminate {
            return None;
        }

        let remaining = match self.s.inner {
            ScannedItem::Rest(s) => &s[self.curr_position..],
            ScannedItem::Str(s) => {
                self.terminate = true;
                return Some(Token::Literal(Literal::String(s)));
            }
            ScannedItem::UnclosedStr(s) => {
                self.terminate = true;
                return Some(Token::Error(LexicalError::UnclosedString(s)));
            }
        };

        if remaining.len() == 0 {
            return None;
        }

        let mut chars = remaining.chars();
        let start_char = chars.next().unwrap();

        // This resolves all unambiguos single chars
        if SINGLE_CHAR_TOKEN.contains_key(&start_char) {
            self.curr_position += 1;
            return SINGLE_CHAR_TOKEN.get(&start_char).cloned();
        }

        // Handle the single unambiguos chars
        if start_char == '<' || start_char == '>' || start_char == ':' || start_char == '=' {
            let next = chars.next();
            let (token, move_forward) = match (start_char, next) {
                ('<', Some('=')) => (Token::Operator(Operator::LesserOrEquals), true),
                ('<', None) | ('<', Some(_)) => (Token::Operator(Operator::LesserThan), false),

                ('>', Some('=')) => (Token::Operator(Operator::GreaterOrEquals), true),
                ('>', None) | ('>', Some(_)) => (Token::Operator(Operator::GreaterThan), false),

                (':', Some('=')) => (Token::Operator(Operator::Assignement), true),
                (':', Some(x)) => (Token::Error(LexicalError::ExpectedAssignement(x)), false),
                (':', None) => (Token::Error(LexicalError::ExpectedAssignementEOF), false),

                ('=', Some('=')) => (Token::Operator(Operator::Equals), true),
                ('=', Some(x)) => (Token::Error(LexicalError::ExpectedDoubleEquals(x)), false),
                ('=', None) => (Token::Error(LexicalError::ExpectedDoubleEqualsEOF), false),

                // If we get here, a lot must be wrong
                _ => return Some(Token::Error(LexicalError::UnknownChar('?'))),
            };

            self.curr_position += 1 + move_forward as usize;
            return Some(token);
        }

        // Get a integer literal
        if start_char.is_numeric() {
            let (token, read) = parse_literal(remaining);
            self.curr_position += read;
            return Some(token);
        }

        // Handle stuff that looks like identifiers
        if start_char.is_alphabetic() || start_char == '_' {
            let (token, read) = parse_identifier(remaining);
            self.curr_position += read;
            return Some(token);
        }

        self.curr_position += 1;
        Some(Token::Error(LexicalError::UnknownChar(start_char)))
    }
}

fn parse_literal(input: &str) -> (Token, usize) {
    // ['1', '0', 'a']
    //             ^
    // ['1', '0']
    // 10
    let first_non_number = input
        .find(|c: char| !c.is_numeric())
        .unwrap_or_else(|| input.len());
    let literal = &input[0..first_non_number];
    (
        Token::Literal(Literal::Integer(literal.parse().unwrap())),
        first_non_number,
    )
}

fn parse_identifier(input: &str) -> (Token, usize) {
    let first_non_id = input
        .find(|c: char| !(c.is_alphanumeric() || c == '_'))
        .unwrap_or_else(|| input.len());
    let id = &input[0..first_non_id];

    if KEYWORDS.contains_key(id) {
        return (KEYWORDS.get(id).cloned().unwrap(), first_non_id);
    }

    (Token::Identifier(id), first_non_id)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::scanner::scan;
    use insta::assert_debug_snapshot;

    #[test]
    fn lex_empty() {
        let mut res = lexicalize(scan(""));
        assert_eq!(res.next(), None);
    }

    #[test]
    fn lex_example() {
        let res: Vec<_> = lexicalize(scan("(x <= > >= < y 99.88l8 )")).collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn lex_real_life() {
        let input = "program fib;\r\nbegin\r\nvar n;\r\nvar first := 0;\r\nvar second :=1;\r\nvar next;\r\nvar c :=0 ;\r\nprint \"enter the number of terms\";\r\nget n;\r\nwhile ( c < n)\r\nbegin\r\nif ( c <= 1)\r\nthen begin next := c; end\r\nelse begin\r\n next := first + second;\r\n second := next;\r\nend\r\nprint next;\r\nc := c + 1;\r\nend\r\nend\r\n";
        let res: Vec<_> = lexicalize(scan(input)).collect();
        assert_debug_snapshot!(res);
    }
}
