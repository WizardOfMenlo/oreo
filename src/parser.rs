use crate::lexical::EnrichedToken;
use crate::parsing_utils::*;
use crate::syntax::*;
use crate::tokens::*;

pub(crate) type ExpToken = Token<'static>;
pub(crate) type TokenList = &'static [ExpToken];

pub enum SyntaxError<'a> {
    ExpectedOneOfButFoundEOF(TokenList),
    ExpectedButFoundEOF(ExpToken),
    ExpectedFound(ExpToken, EnrichedToken<'a>),
    ExpectedOneOfFound(TokenList, EnrichedToken<'a>),
    LogicalError,
}

pub(crate) const ID: ExpToken = Token::Identifier("");

pub fn program<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Program<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Keyword(Keyword::Program))?;
    let id = advance_expecting_identifier(it)?;
    let compound = compound(it)?;

    Ok(Program { id, compound })
}

fn compound<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Compound<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Keyword(Keyword::Begin))?;
    const END: ExpToken = Token::Keyword(Keyword::End);
    let mut statements = Vec::new();
    let mut it = it.peekable();
    loop {
        if let Some(tok) = it.peek() {
            let token = tok.token();
            if token.same_kind(&END) {
                break;
            }

            statements.push(statement(&mut it)?);
        } else {
            return Err(SyntaxError::ExpectedButFoundEOF(END));
        }
    }

    Ok(Compound { statements })
}

fn statement<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Statement<'a>, SyntaxError<'a>> {
    const POSSIBLE_STATEMENT_STARTS: TokenList = &[
        Token::Keyword(Keyword::Var),
        Token::Keyword(Keyword::If),
        Token::Keyword(Keyword::Print),
        Token::Keyword(Keyword::Println),
        Token::Keyword(Keyword::Get),
        Token::Keyword(Keyword::While),
        ID,
    ];

    let matched = advance_expecting_one_of(it, POSSIBLE_STATEMENT_STARTS)?;
    let token = matched.take_token();
    match token {
        Token::Keyword(Keyword::Var) => Ok(Statement::Decl(decl(it)?)),
        Token::Keyword(Keyword::If) => Ok(Statement::If(p_if(it)?)),
        Token::Keyword(Keyword::Print) => Ok(Statement::Print(print(it, token)?)),
        Token::Keyword(Keyword::Println) => Ok(Statement::Print(print(it, token)?)),
        Token::Keyword(Keyword::Get) => Ok(Statement::Print(print(it, token)?)),
        Token::Keyword(Keyword::While) => Ok(Statement::While(p_while(it)?)),
        Token::Identifier(s) => Ok(Statement::Assign(assign(it, Identifier(s))?)),
        _ => Err(SyntaxError::LogicalError),
    }
}

fn decl<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Decl<'a>, SyntaxError<'a>> {
    let id = advance_expecting_identifier(it)?;
    let mut it = it.peekable();
    let next = it.peek();

    match next.map(|s| s.token()) {
        Some(Token::Punctuation(Punctuation::Semicolon)) => Ok(Decl { id, expr: None }),
        _ => Ok(Decl { id, expr: Some(expr(&mut it)?)})
    }
}
 
fn p_if<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<If<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Punctuation(Punctuation::BracketOpen))?;
    let condition = p_bool(it)?;
    advance_expecting(it, Token::Punctuation(Punctuation::BracketClose))?;
    advance_expecting(it, Token::Keyword(Keyword::Then))?;
    let if_branch = compound(it)?;
    let next = advance_expecting_one_of(it, &[Token::Keyword(Keyword::Else), Token::Punctuation(Punctuation::Semicolon)])?;

    let else_branch = match next.token() {
        Token::Keyword(Keyword::Else) => {
            let tmp = compound(it)?;
            advance_expecting(it, Token::Punctuation(Punctuation::Semicolon))?;
            Some(tmp)
        },
        _ => None
    };

    Ok(If {
        condition,
        if_branch,
        else_branch
    })
}

fn print<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
    token: Token<'a>
) -> Result<Print<'a>, SyntaxError<'a>> {
    match token {
        Token::Keyword(Keyword::Print) => Ok(Print::Print(expr(it)?)),
        Token::Keyword(Keyword::Println) => Ok(Print::Println(expr(it)?)),
        Token::Keyword(Keyword::Get) => Ok(Print::Get(advance_expecting_identifier(it)?)),
        // Should only be called with normal stuff
        _ => Err(SyntaxError::LogicalError)
    }
}

fn p_while<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<While<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Punctuation(Punctuation::BracketOpen))?;
    let condition = p_bool(it)?;
    advance_expecting(it, Token::Punctuation(Punctuation::BracketClose))?;
    let compound = compound(it)?;
    advance_expecting(it, Token::Punctuation(Punctuation::Semicolon))?;

    Ok(While {
        condition, 
        compound
    })
}
 
fn assign<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
    id: Identifier<'a>
) -> Result<Assign<'a>, SyntaxError<'a>> {
    advance_expecting(it, Token::Operator(Operator::Assignement))?;
    let expr = expr(it)?;

    Ok(Assign {
        id,
        expr
    })
}
 
fn expr<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Expr<'a>, SyntaxError<'a>> {
    Err(SyntaxError::LogicalError)
}

fn p_bool<'a>(
    it: &mut impl Iterator<Item = EnrichedToken<'a>>,
) -> Result<Bool<'a>, SyntaxError<'a>> {
    Err(SyntaxError::LogicalError)
}