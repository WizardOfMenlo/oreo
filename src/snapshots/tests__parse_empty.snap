---
source: src/parser.rs
expression: parsed
---
Node {
    ty: Error(
        ExpectedFoundEOF(
            Keyword(
                Program,
            ),
        ),
    ),
    text_range: 0..0,
    children: [],
}
